ledger/src/snapshot_utils.rs

```diff
@@ -349,6 +418,93 @@ where
     Ok(ret)
 }
 
+#[allow(clippy::type_complexity)]
+fn create_versioned_bank_snapshot_serializer<'a>(
+    snapshot_version: SnapshotVersion,
+    bank: &'a Bank,
+    snapshot_storages: &'a [SnapshotStorage],
+) -> Result<Box<dyn Fn(&mut BufWriter<File>) -> Result<()> + 'a>> {
+    match snapshot_version {
+        SnapshotVersion::V1_1_0 => Ok(Box::new(
+            move |stream: &mut BufWriter<File>| -> Result<()> {
+                serialize_into(stream.by_ref(), bank)?;
+                context_bankrc_to_stream::<SerdeContextV1_1_0, _>(
+                    stream.by_ref(),
+                    &bank.rc,
+                    snapshot_storages,
+                )?;
+                Ok(())
+            },
+        )),
+        SnapshotVersion::V1_1_1 => Ok(Box::new(
+            move |stream: &mut BufWriter<File>| -> Result<()> {
+                serialize_into(stream.by_ref(), bank)?;
+                context_bankrc_to_stream::<SerdeContextV1_1_1, _>(
+                    stream.by_ref(),
+                    &bank.rc,
+                    snapshot_storages,
+                )?;
+                Ok(())
+            },
+        )),
+        //If additional snapshot versions were defined but not implemented...
+        //_ => Err(get_io_error(&format!("unsupported snapshot version: {}", snapshot_version.to_string()))),
+    }
+}
+
+#[allow(clippy::type_complexity)]
+fn create_versioned_bank_snapshot_deserializer<'a, P: AsRef<Path>>(
+    snapshot_version: &str,
+    account_paths: &'a [PathBuf],
+    frozen_account_pubkeys: &'a [Pubkey],
+    append_vecs_path: &'a P,
+    genesis_config: &GenesisConfig,
+) -> Result<Box<dyn Fn(&mut BufReader<File>) -> Result<Bank> + 'a>> {
+    let operating_mode = genesis_config.operating_mode;
+    match SnapshotVersion::maybe_from_string(snapshot_version) {
+        Some(SnapshotVersion::V1_1_0) => Ok(Box::new(move |stream: &mut BufReader<File>| {
+            let mut bank: Bank = bincode::config()
+                .limit(MAX_SNAPSHOT_DATA_FILE_SIZE)
+                .deserialize_from(stream.by_ref())?;
+            bank.operating_mode = Some(operating_mode);
+            info!("Rebuilding accounts...");
+            let rc = context_bankrc_from_stream::<SerdeContextV1_1_0, _, _>(
+                account_paths,
+                bank.slot(),
+                &bank.ancestors,
+                frozen_account_pubkeys,
+                stream.by_ref(),
+                &append_vecs_path,
+            )?;
+            bank.rc = rc;
+            bank.finish_init();
```

Comment:
nits: strictly speaking, this isn't related to (versioned) (de-)serialization, so it could be outside this `match`.

Reply:
Strictly speaking the function is intended to return an initialized bank hence the rc assignment and finish_init().

It really should be done in a ctor but the bank currently comes first in the stream which is unfortunate; the needless deafult creation of a new BankRc / Accounts etc. is painful.

Of course the presence of numerous #[serde(skip)] attributed fields in Bank is a sign that it shouldn't be serialised directly itself. Indeed I was curious how the message_processor in Bank was serialised and of course we have:
```
#[derive(Default, Deserialize, Serialize)]
pub struct MessageProcessor {
    #[serde(skip)]
    instruction_processors: Vec<(Pubkey, ProcessInstruction)>,
    #[serde(skip)]
    native_loader: NativeLoader,
}
```

Reply:
I've pulled the generate_index / freeze_accounts calls up into rebuild_bank_from_snapshots().
freeze_accounts in particular takes a &mut self which requires nested Arc::get_mut() which are somewhat unsightly but c'est la vie.

Reply:
(Delayed commenting) - I put generate_index() back as a number of testcases expected accounts_from_stream & accountsdb_from_stream to return a usable AccountsDB (i.e. with the index rebuilt).

Reply:
Ok; thanks for trying and thinking about these! Then, let's settle on this as-is. :)

Reply:
Come to think of it, due to rust's life cycle, maybe my separation-of-concerns thinking should be wrapped around it. :) Sorry for my premature commenting!